server/src/server.rs

```diff
@@ -201,32 +193,38 @@ impl<M: ConnectionManager> Server<M> {
     pub async fn write_lines(&self, db_name: &str, lines: &[ParsedLine<'_>]) -> Result<()> {
         let id = self.require_id()?;
 
+        let db_name = DatabaseName::new(db_name).context(InvalidDatabaseName)?;
         let db = self
             .config
             .databases
-            .get(db_name)
-            .context(DatabaseNotFound { db_name })?;
+            .get(&db_name)
+            .context(DatabaseNotFound {
+                db_name: db_name.to_string(),
```

Comment:
This will allocate a new string whether or not the error happens. `SNAFU`'s `context` calls `into` in a closure, so it's better to pass references here. Rust needs a little help figuring out the deref coercion; this works for me: 

```suggestion
                db_name: &*db_name,
```

Reply:
Ah nice - the closure thing is pretty neat!